pyomo/contrib/sensitivity_toolbox/k_aug.py

```diff
@@ -0,0 +1,166 @@
+# ______________________________________________________________________________
+#
+# Pyomo: Python Optimization Modeling Objects
+# Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+# Under the terms of Contract DE-NA0003525 with National Technology and
+# Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+# rights in this software.
+# This software is distributed under the 3-clause BSD License
+# ______________________________________________________________________________
+import os
+import enum
+from pyomo.environ import (
+        SolverFactory,
+        )
+from pyomo.common.tempfiles import TempfileManager
+
+
+class FileType(enum.Enum):
+    K_AUG_INPUT = 1
+    K_AUG_OUTPUT = 2
+    DOT_SENS_INPUT = 3
+    DOT_SENS_OUTPUT = 4
+FT = FileType
+
+
+def _attribute_from_filename(filename):
+    base_name = os.path.basename(filename)
+    return base_name.split(".")[0]
+
+
+debug_dir = "kaug_debug"
+# These are files we would like to save from a call to k_aug
+# or dot_sens. Other files generated will still be deleted,
+# but not saved on the K_augInterface object.
+known_files = {
+        "dsdp_in_.in": {FT.K_AUG_OUTPUT, FT.DOT_SENS_INPUT},
+        "conorder.txt": {FT.K_AUG_OUTPUT},
+        "timings_k_aug_dsdp.txt": {FT.K_AUG_OUTPUT},
+        "col_row.nl": {FT.K_AUG_OUTPUT},
+        "col_row.col": {FT.K_AUG_OUTPUT},
+        "col_row.row": {FT.K_AUG_OUTPUT},
+        os.path.join(debug_dir, "kkt.in"): {FT.K_AUG_OUTPUT},
+        "dot_out.out": {FT.DOT_SENS_OUTPUT},
+        "delta_p.out": {FT.DOT_SENS_OUTPUT},
+        "timings_dot_driver_dsdp.txt": {FT.DOT_SENS_OUTPUT},
+        }
+
+file_attr_map = {name: _attribute_from_filename(name) for name in known_files}
+
+k_aug_input_files = [name for name, types in known_files.items()
+        if FT.K_AUG_INPUT in types]
+k_aug_output_files = [name for name, types in known_files.items()
+        if FT.K_AUG_OUTPUT in types]
+dot_sens_input_files = [name for name, types in known_files.items()
+        if FT.DOT_SENS_INPUT in types]
+dot_sens_output_files = [name for name, types in known_files.items()
+        if FT.DOT_SENS_OUTPUT in types]
+
+
+class K_augInterface(object):
+    """
+    k_aug and dot_sens store information in the user's filesystem,
+    some of which is mandatory for subsequent calls.
+    This class ensures that calls to these executables happen in 
+    temporary directories. The resulting files are immediately read
+    and cached as attributes of this object, and the temporary
+    directories deleted. If we have cached files that can be used
+    by a subsequent call to k_aug or dot_sens, we write them just
+    before calling the executable, and they are deleted along with
+    the temporary directory.
+
+    NOTE: only covers dsdp_mode for now.
+    """
```

Comment:
Created a `K_augInterface` class to cache the files k_aug writes, and call k_aug/dot_sens inside of a temporary directory. Not too thrilled with how I'm storing the data right now, but this should let us avoid populating the user's working directory.

Reply:
Cleaned up this this implementation a bit. This PR still needs changes to doctests and to fix typos.